node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+    /// Location of the protobuf_build crate, visible from the generated code.
+    fn this_crate(&self) -> RustName {
+        self.protobuf_crate.clone().join("build")
+    }
+
+    /// Generates implementation of `prost_reflect::ReflectMessage` for a rust type generated
+    /// from a message of the given `proto_name`.
+    fn reflect_impl(&self, proto_name: ProtoName) -> String {
+        let rust_name = proto_name
+            .relative_to(&self.proto_root.to_name())
+            .unwrap()
+            .to_rust_type()
+            .to_string();
+        let proto_name = proto_name.to_string();
+        let this = self.this_crate().to_string();
+        format!("impl {this}::prost_reflect::ReflectMessage for {rust_name} {{\
+            fn descriptor(&self) -> {this}::prost_reflect::MessageDescriptor {{\
+                static INIT : {this}::Lazy<{this}::prost_reflect::MessageDescriptor> = {this}::Lazy::new(|| {{\
+                    DESCRIPTOR.load_global().get_message_by_name({proto_name:?}).unwrap()\
+                }});\
+                INIT.clone()\
+            }}\
+        }}")
+    }
+
+    /// Generates rust code from the proto files according to the config.
+    pub fn generate(&self) -> anyhow::Result<()> {
+        assert!(
+            self.input_root.abs().is_dir(),
+            "input_root should be a directory"
+        );
+        println!("cargo:rerun-if-changed={}", self.input_root.to_str());
+
+        // Load dependencies.
+        let mut pool = prost_reflect::DescriptorPool::new();
+        for d in &self.dependencies {
+            d.1.load(&mut pool)
+                .with_context(|| format!("failed to load dependency {}", d.0))?;
+        }
+        let mut pool_raw = prost_types::FileDescriptorSet::default();
+        pool_raw.file.extend(pool.file_descriptor_protos().cloned());
+
+        // Load proto files.
+        let mut proto_paths = vec![];
+        traverse_files(&self.input_root.abs(), &mut |path| {
+            let Some(ext) = path.extension() else {
+                return Ok(());
+            };
+            let Some(ext) = ext.to_str() else {
+                return Ok(());
+            };
+            if ext != "proto" {
+                return Ok(());
+            };
+
+            let file_raw = fs::read_to_string(path).context("fs::read()")?;
+            let path = ProtoPath::from_input_path(path, &self.input_root, &self.proto_root);
+            pool_raw
+                .file
+                .push(protox_parse::parse(path.to_str(), &file_raw).map_err(
+                    // rewrapping the error, so that source location is included in the error message.
+                    |err| anyhow::anyhow!("{:?}", err),
+                )?);
+            proto_paths.push(path);
+            Ok(())
+        })?;
+
+        // Compile the proto files
+        let mut compiler = protox::Compiler::with_file_resolver(
+            protox::file::DescriptorSetFileResolver::new(pool_raw),
+        );
```

Comment:
Stupid question: is this separate compiler (as opposed to `prost_build`) used just to create a file descriptor set? Does `prost_build` lack the necessary functionality to do so? It allows [to output FDS at the specified path](https://docs.rs/prost-build/latest/prost_build/struct.Config.html#method.file_descriptor_set_path).

Reply:
prost_build doesn't include the compiler. It proxies the compilation to a protoc compiler which has to be included in the system or be imported as a binary. This compiler is more lightweight and more straightforward to instrument.